server/src/main/java/org/apache/kafka/server/LogReadResult.java

```diff
@@ -0,0 +1,129 @@
+) {
+    public LogReadResult(
+            FetchDataInfo info,
+            Optional<FetchResponseData.EpochEndOffset> divergingEpoch,
+            long highWatermark,
+            long leaderLogStartOffset,
+            long leaderLogEndOffset,
+            long followerLogStartOffset,
+            long fetchTimeMs,
+            OptionalLong lastStableOffset) {
+        this(info, divergingEpoch, highWatermark, leaderLogStartOffset, leaderLogEndOffset, followerLogStartOffset,
+            fetchTimeMs, lastStableOffset, OptionalInt.empty(), Optional.empty());
+    }
+
+    public LogReadResult(
+        FetchDataInfo info,
+        Optional<FetchResponseData.EpochEndOffset> divergingEpoch,
+        long highWatermark,
+        long leaderLogStartOffset,
+        long leaderLogEndOffset,
+        long followerLogStartOffset,
+        long fetchTimeMs,
+        OptionalLong lastStableOffset,
+        Optional<Throwable> exception) {
+        this(info, divergingEpoch, highWatermark, leaderLogStartOffset, leaderLogEndOffset, followerLogStartOffset,
+            fetchTimeMs, lastStableOffset, OptionalInt.empty(), exception);
+    }
+
+    public LogReadResult(
+            FetchDataInfo info,
+            Optional<FetchResponseData.EpochEndOffset> divergingEpoch,
+            long highWatermark,
+            long leaderLogStartOffset,
+            long leaderLogEndOffset,
+            long followerLogStartOffset,
+            long fetchTimeMs,
+            OptionalLong lastStableOffset,
+            OptionalInt preferredReadReplica) {
+        this(info, divergingEpoch, highWatermark, leaderLogStartOffset, leaderLogEndOffset, followerLogStartOffset,
+            fetchTimeMs, lastStableOffset, preferredReadReplica, Optional.empty());
+    }
+
+    public Errors error() {
+        if (exception.isPresent()) {
+            return Errors.forException(exception.get());
+        }
+        return Errors.NONE;
+    }
+
+    @Override
+    public String toString() {
+        return "LogReadResult(info=" + info +
+               ", divergingEpoch=" + divergingEpoch +
+               ", highWatermark=" + highWatermark +
+               ", leaderLogStartOffset" + leaderLogStartOffset +
+               ", leaderLogEndOffset" + leaderLogEndOffset +
+               ", followerLogStartOffset" + followerLogStartOffset +
+               ", fetchTimeMs=" + fetchTimeMs +
+               ", preferredReadReplica=" + preferredReadReplica +
+               ", lastStableOffset=" + lastStableOffset +
+               ", error=" + error() + ")";
+    }
+
+    public FetchPartitionData toFetchPartitionData(boolean isReassignmentFetch) {
+        return new FetchPartitionData(
+            this.error(),
+            this.highWatermark,
+            this.leaderLogStartOffset,
+            this.info.records,
+            this.divergingEpoch,
+            this.lastStableOffset.isPresent() ? OptionalLong.of(this.lastStableOffset.getAsLong()) : OptionalLong.empty(),
```

Comment:
Why not just simply return the value? Earlier in scala it was Option to Optional... conversion, seems not required now.

```suggestion
            this.lastStableOffset,
```